packages/zoe/src/contracts/multipoolAutoswap/getCurrentPrice.js

```diff
@@ -55,30 +64,129 @@ export const makeGetCurrentPrice = (isSecondary, isCentral, getPool) => {
+      // 1) from amountIn to the central token, which tells us how much central
+      // would be provided for amountIn (centralAmount)
+      // 2) from centralAmount to brandOut, which tells us how much of brandOut
+      // will be provided (amountOut) as well as the minimum price in central
+      // tokens (reducedCentralAmount), then finally
+      // 3) call getPriceGivenRequiredOutput() to see if the same proceeds can
+      // be purchased for less (reducedAmountIn).
+
+      const brandInPool = getPool(brandIn);
+      const brandOutPool = getPool(brandOut);
+      const {
+        amountOut: centralAmount,
+      } = brandInPool.getPriceGivenAvailableInput(amountIn, centralBrand);
+      const {
+        amountIn: reducedCentralAmount,
+        amountOut,
+      } = brandOutPool.getPriceGivenAvailableInput(centralAmount, brandOut);
+
+      // propagate reduced prices back to the first pool
+      const {
+        amountIn: reducedAmountIn,
+      } = brandInPool.getPriceGivenRequiredOutput(
         brandIn,
-      ).getSecondaryToCentralOutputPrice(outputValue);
-      return getPool(brandOut).getCentralToSecondaryOutputPrice(
-        centralTokenAmount.value,
+        reducedCentralAmount,
       );
+      return {
+        amountIn: reducedAmountIn,
+        amountOut,
+        centralAmount: reducedCentralAmount,
+      };
     }
 
     throw new Error(`brands were not recognized`);
   };
 
-  return { getOutputForGivenInput, getInputForGivenOutput };
+  const getPriceGivenRequiredOutput = (brandIn, amountOut) => {
+    const { brand: brandOut } = amountOut;
+
+    if (isCentral(brandIn) && isSecondary(brandOut)) {
+      return getPool(brandOut).getPriceGivenRequiredOutput(brandIn, amountOut);
+    } else if (isSecondary(brandIn) && isCentral(brandOut)) {
+      return getPool(brandIn).getPriceGivenRequiredOutput(brandIn, amountOut);
+    } else if (isSecondary(brandIn) && isSecondary(brandOut)) {
+      // We must do two consecutive getPriceGivenRequiredOutput() calls,
+      // followed by a call to getPriceGivenAvailableInput().
+      // 1) from amountOut to the central token, which tells us how much central
+      // is required to obtain amountOut (centralAmount)
+      // 2) from centralAmount to brandIn, which tells us how much of brandIn
+      // is required (amountIn) as well as the max proceeds in central
+      // tokens (improvedCentralAmount), then finally
+      // 3) call getPriceGivenAvailableInput() to see if improvedCentralAmount
+      // produces a larger amount (improvedAmountOut)
+
+      const brandInPool = getPool(brandOut);
+      const brandOutPool = getPool(brandIn);
+
+      const {
+        amountIn: centralAmount,
+      } = brandInPool.getPriceGivenRequiredOutput(centralBrand, amountOut);
+      const {
+        amountIn,
+        amountOut: improvedCentralAmount,
+      } = brandOutPool.getPriceGivenRequiredOutput(brandIn, centralAmount);
+
```

Comment:
This logic is the correct version, whereas `getInputForGivenOutput` has incorrect logic. I think for this reason, we should never have this logic in duplicate. It's too easy to get wrong.

Reply:
This version turns out to have been subtly wrong. I thought I could get away without testing every single case, but that was wrong.